# A Short and Concise Python Tutorial 

### Instructor: Dr. M. Nedim ALPDEMİR

*This material is prepared as part of the EE592-Advanced Modeling And Simulation course given at Yıldırım Beyazıd University (YBÜ).*

 * Python is a cross-platform programming language, meaning, it runs on multiple platforms like Windows, Mac OS X, Linux, Unix and has even been ported to the Java and .NET virtual machines. It is free and open source.
 
 ## There are 3 ways of writing and executing python code
 
### 1. Immediate mode
 
Typing python in the command line will invoke the interpreter in immediate mode. We can directly type in Python expressions and press enter to get the output.

`>>>` is the Python prompt. It tells us that the interpreter is ready for our input. 

### 2. Script mode
This mode is used to execute Python program written in a file. Such a file is called a script. Scripts can be saved to disk for future use. Python scripts have the extension .py, meaning that the filename ends with .py.

For example: `helloWorld.py`

To execute this file in script mode we simply write `python helloWorld.py` at the command prompt.

### 3. Integrated Development Environment (IDE)

We can use any text editing software to write a Python script file.

We just need to save it with the .py extension. But using an IDE can make our life a lot easier. IDE is a piece of software that provides useful features like code hinting, syntax highlighting and checking, file explorers etc. to the programmer for application development.

### Python Statements

Instructions that a Python interpreter can execute are called statements.

* multi line / single line statements
* comments
* the importance of indentation

In [3]:
# single line comment

'''
multi
line
comment
'''
# a multi line statement
a = 1 + 2 + 3 + \
    4 + 5 + 6 + \
    7 + 8 + 9

# This is explicit line continuation. 
# In Python, line continuation is implied inside parentheses ( ), brackets [ ] and braces { }.

colors = ['red',
          'blue',
          'green']

# indentation is very important
'''
A code block (body of a function, loop etc.) starts with indentation and ends with the first unindented line. 
The amount of indentation is up to you, but it must be consistent throughout that block.
'''

for i in range(1,11):
 print(i)
     if i == 5:
        break

IndentationError: unexpected indent (<ipython-input-3-68225b9a7b6a>, line 26)

# Python Variables, Constants and Literals

In Python, variables do not need declaration to reserve memory space. The "variable declaration" or "variable initialization" happens automatically when we assign a value to a variable.

In [5]:
### website = "Apple.com"
print ('website')

# assigning to multiple variables in ne statement
a, b, c = 5, 3.2, "Hello"
# or similarly
x = y = z = "same"

print (a)
print (b)
print (c)

website
5
3.2
Hello


In [ ]:
# constants
PI = 3.14
GRAVITY = 9.8

# can be used diectly
area = PI*3**2
print(area)


# or can be used by importing an external file where those constants were defined 
import MyConstants
newArea = MyConstants.MY_PI*5**2
print(newArea)

In [6]:
# Numeric literals

a = 0b1010 #Binary Literals
b = 100 #Decimal Literal 
c = 0o310 #Octal Literal
d = 0x12c #Hexadecimal Literal

#Float Literal
float_1 = 10.5 
float_2 = 1.5e2

#Complex Literal 
x = 3.14j

print(a, b, c, d)
print ("{:b}".format(a))
print(float_1, float_2)
print(x, x.imag, x.real)

10 100 200 300
1010
10.5 150.0
3.14j 3.14 0.0


In [7]:
# Boolean liteals
x = (1 == True)
y = (1 == False)
a = True + 4
b = False + 10

print("x is", x)
print("y is", y)
print("a:", a)
print("b:", b)

x is True
y is False
a: 5
b: 10


In [10]:
# liteal collections
fruits = ["apple", "mango", "orange"] #list
numbers = (1, 2, 3) #tuple
alphabets = {'a':'apple', 'b':'ball', 'c':'cat'} #dictionary
vowels = {'a', 'e', 'i' , 'o', 'u'} #set

print(fruits)
print(numbers)
print(alphabets)
print(vowels)

print(alphabets['a'])

['apple', 'mango', 'orange']
(1, 2, 3)
{'a': 'apple', 'b': 'ball', 'c': 'cat'}
{'e', 'a', 'i', 'o', 'u'}
apple


# Python Data Types

## Lists

List is an ordered sequence of items. It is one of the most used datatype in Python and is very flexible. All the items in a list do not need to be of the same type.

Declaring a list is pretty straight forward. Items separated by commas are enclosed within brackets [ ].


In [11]:
a = [5,10,15,20,25,30,35,40]

# print the third element, a[2] = 15
print("a[2] = ", a[2])

# print the first three element a[0:3] = [5, 10, 15]
print("a[0:3] = ", a[0:3])

# print all the elements after the 5th element including the 5th, a[5:] = [30, 35, 40]
print("a[5:] = ", a[5:])

# append a new element
a.append(100)
print(a)

# extend the list with new elements
a.extend([110,120,130])
print(a)

#replace an element with a new value
a[3] = 250
print(a)

a[2] =  15
a[0:3] =  [5, 10, 15]
a[5:] =  [30, 35, 40]
[5, 10, 15, 20, 25, 30, 35, 40, 100]
[5, 10, 15, 20, 25, 30, 35, 40, 100, 110, 120, 130]
[5, 10, 15, 250, 25, 30, 35, 40, 100, 110, 120, 130]


# Tuples

* **Tuple** is an ordered sequence of items same as list.The only difference is that tuples are immutable. 
* Tuples once created cannot be modified.
* Tuples are used to write-protect data and are usually faster than list as it cannot change dynamically.
* It is defined within parentheses () where items are separated by commas.

In [12]:
t = (5,'program', 1+3j)

# t[1] = 'program'
print("t[1] = ", t[1])

# t[0:3] = (5, 'program', (1+3j))
print("t[0:3] = ", t[0:3])

# Generates error
# Tuples are immutable
t[0] = 10

t[1] =  program
t[0:3] =  (5, 'program', (1+3j))


TypeError: 'tuple' object does not support item assignment

# Python Strings

* String is sequence of Unicode characters. We can use single quotes or double quotes to represent strings. 
* Multi-line strings can be denoted using triple quotes, ''' or """.

In [13]:
s = 'Hello world!'

# s[4] = 'o'
print("s[4] = ", s[4])

# s[6:11] = 'world'
print("s[6:11] = ", s[6:11])

# Generates error
# Strings are immutable in Python
s[5] ='d'


s[4] =  o
s[6:11] =  world


TypeError: 'str' object does not support item assignment

# Python Set

* Set is an unordered collection of unique items. 
* Set is defined by values separated by comma inside braces { }. 
* Items in a set are not ordered.
* Items are not indexed so they can not be accessed via index operator []


In [14]:
a = {5,2,3,1,4}

# printing set variable
print("a = ", a)

# data type of variable a
print(type(a))

# sets elements are unique so even if you include multiples it will eliminate them
b = {1,1,2,2,3,3,3,6,8}
print("b = ", b)

# you can apply set operations such as UNION, INTERSECTION, CARTESIAN PRODUCT, DIFFERENCE etc.
# UNION
c = a | b
print("a | b = ", c)

#INTERSECTION
c = a & b
print("a & b = ", c)

#DIFFERENCE
c = a - b
print("a - b = ", c)

#for cartesian product you need to use a library function
from itertools import product
c = product(a,b)
#the prodyct itself is an object
print(c)

#you can access the elements of the cartesian product in a loop
for e in c:
    print(e)
    
# or construct a list of tuples from the cartesian products
tuple_list = [e for e in c]
print(tuple_list)
#print(e for e in c)

a =  {1, 2, 3, 4, 5}
<class 'set'>
b =  {1, 2, 3, 6, 8}
a | b =  {1, 2, 3, 4, 5, 6, 8}
a & b =  {1, 2, 3}
a - b =  {4, 5}
(1, 1)
(1, 2)
(1, 3)
(1, 6)
(1, 8)
(2, 1)
(2, 2)
(2, 3)
(2, 6)
(2, 8)
(3, 1)
(3, 2)
(3, 3)
(3, 6)
(3, 8)
(4, 1)
(4, 2)
(4, 3)
(4, 6)
(4, 8)
(5, 1)
(5, 2)
(5, 3)
(5, 6)
(5, 8)
[]


# Python Dictionary

* **Dictionary** is an unordered collection of key-value pairs.
* It is generally used when we have a huge amount of data. 
* Dictionaries are optimized for retrieving data. We must know the key to retrieve the value.
* In Python, dictionaries are defined within braces `{}` with each item being a pair in the form `key:value`. Key and value can be of any type.

In [15]:
d = {1:'value','key':2}
print(type(d))

print("d[1] = ", d[1]);

print("d['key'] = ", d['key']);

# nested dictionaries and lists are possible
person_info = {'name':"Ayşe", 'age':24, 'hobbies':["reading","running","coding"], 
        'phones':{'home':"0234 55 66", 'work':"7766876",'mobile':"0546 77 8777"}}

print("name : ",person_info['name'])
print("hobbies : ",person_info['hobbies'])
print("mobile phone : ",person_info['phones']['mobile'])

# Generates error
#print("d[2] = ", d[2]);


<class 'dict'>
d[1] =  value
d['key'] =  2
name :  Ayşe
hobbies :  ['reading', 'running', 'coding']
mobile phone :  0546 77 8777


In [16]:
'''
CONVERSION from one data type to another is easy
'''

# convert list to set : note that the multiples are eliminated
s = set([1,2,3,3,4,4])
print(s)

#convert set to tuple
t = tuple({5,6,7})
print(t)
# convert string to list
str_list = list('hello')
print(str_list)

# convert list of lists to dictionary
d1 =  dict([[1,2],[3,4]])
print(d1)

# convert list of tuples  to dictionary
d2 = dict([(3,26),(4,44)])
print(d2)

{1, 2, 3, 4}
(5, 6, 7)
['h', 'e', 'l', 'l', 'o']
{1: 2, 3: 4}
{3: 26, 4: 44}


<h1><a id="arithmetic" name="arithmetic"></a>Arithmetic operators</h1>

<p>Arithmetic operators are used to perform mathematical operations like addition, subtraction, multiplication etc.</p>

<table border="1">
	<caption>Arithmetic operators in Python</caption>
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Meaning</th>
			<th>Example</th>
		</tr>
		<tr>
			<td>+</td>
			<td>Add two operands or unary plus</td>
			<td>x + y<br />
				+2</td>
		</tr>
		<tr>
			<td>-</td>
			<td>Subtract right operand from the left or unary minus</td>
			<td>x - y<br />
				-2</td>
		</tr>
		<tr>
			<td>*</td>
			<td>Multiply two operands</td>
			<td>x * y</td>
		</tr>
		<tr>
			<td>/</td>
			<td>Divide left operand by the right one (always results into float)</td>
			<td>x / y</td>
		</tr>
		<tr>
			<td>%</td>
			<td>Modulus - remainder of the division of left operand by the right</td>
			<td>x % y (remainder of x/y)</td>
		</tr>
		<tr>
			<td>//</td>
			<td>Floor division - division that results into whole number adjusted to the left in the number line</td>
			<td>x // y</td>
		</tr>
		<tr>
			<td>**</td>
			<td>Exponent - left operand raised to the power of right</td>
			<td>x**y (x to the power y)</td>
		</tr>
	</tbody>
</table>

# Comparison Operators

<table border="1">
	<caption>Comparision operators in Python</caption>
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Meaning</th>
			<th>Example</th>
		</tr>
		<tr>
			<td>&gt;</td>
			<td>Greater that - True if left operand is greater than the right</td>
			<td>x &gt; y</td>
		</tr>
		<tr>
			<td>&lt;</td>
			<td>Less that - True if left operand is less than the right</td>
			<td>x &lt; y</td>
		</tr>
		<tr>
			<td>==</td>
			<td>Equal to - True if both operands are equal</td>
			<td>x == y</td>
		</tr>
		<tr>
			<td>!=</td>
			<td>Not equal to - True if operands are not equal</td>
			<td>x != y</td>
		</tr>
		<tr>
			<td>&gt;=</td>
			<td>Greater than or equal to - True if left operand is greater than or equal to the right</td>
			<td>x &gt;= y</td>
		</tr>
		<tr>
			<td>&lt;=</td>
			<td>Less than or equal to - True if left operand is less than or equal to the right</td>
			<td>x &lt;= y</td>
		</tr>
	</tbody>
</table>

# Logical operators

<p>Logical operators are the <code>and</code>, <code>or</code>, <code>not</code> operators.</p>

<table border="1">
	<caption>Logical operators in Python</caption>
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Meaning</th>
			<th>Example</th>
		</tr>
		<tr>
			<td>and</td>
			<td>True if both the operands are true</td>
			<td>x and y</td>
		</tr>
		<tr>
			<td>or</td>
			<td>True if either of the operands is true</td>
			<td>x or y</td>
		</tr>
		<tr>
			<td>not</td>
			<td>True if operand is false (complements the operand)</td>
			<td>not x</td>
		</tr>
	</tbody>
</table>


<h1 id="bitwise_operators"><a id="bitwise" name="bitwise"></a>Bitwise operators</h1>

<p>Bitwise operators act on operands as if they were string of binary digits. It operates bit by bit, hence the name.</p>

<p>For example, 2 is <code>10</code> in binary and 7 is <code>111</code>.</p>

<p><strong>In the table below:</strong> Let <var>x</var> = 10 (<code>0000 1010</code> in binary) and <var>y</var> = 4 (<code>0000 0100</code> in binary)</p>

<table border="1">
	<caption>Bitwise operators in Python</caption>
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Meaning</th>
			<th>Example</th>
		</tr>
		<tr>
			<td>&amp;</td>
			<td>Bitwise AND</td>
			<td>x&amp; y = 0 (<code>0000 0000</code>)</td>
		</tr>
		<tr>
			<td>|</td>
			<td>Bitwise OR</td>
			<td>x | y = 14 (<code>0000 1110</code>)</td>
		</tr>
		<tr>
			<td>~</td>
			<td>Bitwise NOT</td>
			<td>~x = -11 (<code>1111 0101</code>)</td>
		</tr>
		<tr>
			<td>^</td>
			<td>Bitwise XOR</td>
			<td>x ^ y = 14 (<code>0000 1110</code>)</td>
		</tr>
		<tr>
			<td>&gt;&gt;</td>
			<td>Bitwise right shift</td>
			<td>x&gt;&gt; 2 = 2 (<code>0000 0010</code>)</td>
		</tr>
		<tr>
			<td>&lt;&lt;</td>
			<td>Bitwise left shift</td>
			<td>x&lt;&lt; 2 = 40 (<code>0010 1000</code>)</td>
		</tr>
	</tbody>
</table>



<h1><a id="membership" name="membership"></a>Membership operators</h1>

<p><code>in</code> and <code>not in</code> are the membership operators in Python. They are used to test whether a value or variable is found in a sequence (<a href="/python-programming/string" title="Python strings">string</a>, <a href="/python-programming/list" title="Python list">list</a>, <a href="/python-programming/tuple" title="Python tuple">tuple</a>, <a href="/python-programming/set" title="Python set">set</a> and <a href="/python-programming/dictionary" title="Python dictionary">dictionary</a>).</p>

<p>In a dictionary we can only test for presence of key, not the value.</p>

<table border="1">
	<tbody>
		<tr>
			<th>Operator</th>
			<th>Meaning</th>
			<th>Example</th>
		</tr>
		<tr>
			<td>in</td>
			<td>True if value/variable is found in the sequence</td>
			<td>5 in x</td>
		</tr>
		<tr>
			<td>not in</td>
			<td>True if value/variable is not found in the sequence</td>
			<td>5 not in x</td>
		</tr>
	</tbody>
</table>

# Output formatting

Sometimes we would like to format our output to make it look attractive. This can be done by using the str.format() method. This method is visible to any string object.

In [17]:
print('I love {0} and {1}'.format('bread','butter'))
# Output: I love bread and butter

print('I love {1} and {0}'.format('bread','butter'))
# Output: I love butter and bread

print('Hello {name}, {greeting}'.format(greeting = 'Goodmorning', name = 'John'))

x = 12.3456789
print('The value of x is %3.2f' %x)
print('The value of x is %3.4f' %x)

I love bread and butter
I love butter and bread
Hello John, Goodmorning
The value of x is 12.35
The value of x is 12.3457


# Python Input

To allow flexibility we might want to take the input from the user. In Python, we have the input() function to allow this. The syntax for input() is

`input([prompt])`


In [18]:
num = input('Enter a number: ')
print("you entered = ",num)

# input gets the user input as a string, so cast it to a number if you want to use it in a calculation
sum = 23 + float(num)
print("total = ",sum)

Enter a number: 23
you entered =  23
total =  46.0


# Python if...elif...else Statement

flow diagram is as follows:

![Python_if_elif_else_statement.jpg](Python_if_elif_else_statement.jpg)

In [19]:
# In this program, 
# we check if the number is positive or
# negative or zero and 
# display an appropriate message

num = 3.4

# Try these two variations as well:
# num = 0
# num = -4.5

if num > 0:
    print("Positive number")
elif num == 0:
    print("Zero")
else:
    print("Negative number")
    
    
# nested if statements

# In this program, we input a number
# check if the number is positive or
# negative or zero and display
# an appropriate message
# This time we use nested if

num = float(input("Enter a number: "))
if num >= 0:
    if num == 0:
        print("Zero")
    else:
        print("Positive number")
else:
    print("Negative number")

Positive number
Enter a number: 23
Positive number


# For Loop in Python

he for loop in Python is used to iterate over a sequence (list, tuple, string) or other iterable objects. Iterating over a sequence is called traversal.

Syntax of for Loop

`
for val in sequence:
	Body of for
`

Here, val is the variable that takes the value of the item inside the sequence on each iteration.

Loop continues until we reach the last item in the sequence. The body of for loop is separated from the rest of the code using indentation.

## The range() function

* We can generate a sequence of numbers using `range()` function. `range(10)` will generate numbers from 0 to 9 (10 numbers).

* We can also define the start, stop and step size as `range(start,stop,step size)`. step size defaults to 1 if not provided.

* This function does not store all the values in memory, it would be inefficient. So it remembers the start, stop, step size and generates the next number on the go.

* To force this function to output all the items, we can use the function list()


In [20]:
# Program to find the sum of all numbers stored in a list

# List of numbers
numbers = [6, 5, 3, 8, 4, 2, 5, 4, 11]

# variable to store the sum
sum = 0

# iterate over the list
for val in numbers:
	sum = sum+val

# Output: The sum is 48
print("The sum is", sum)

The sum is 48


In [21]:
# Output: range(0, 10)
print(range(10))

# Output: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
print(list(range(10)))

# Output: [2, 3, 4, 5, 6, 7]
print(list(range(2, 8)))

# Output: [2, 5, 8, 11, 14, 17]
print(list(range(2, 20, 3)))

range(0, 10)
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[2, 3, 4, 5, 6, 7]
[2, 5, 8, 11, 14, 17]


In [22]:
# Program to iterate through a list using indexing

genre = ['nature', 'jogging', 'music']

# iterate over the list using index
for i in range(len(genre)):
	print("I like", genre[i])

I like nature
I like jogging
I like music


## for loop with else
* A for loop can have an optional else block as well. The `else` part is executed if the items in the sequence used in for loop exhausts.

* break statement can be used to stop a for loop. In such case, the else part is ignored.

* Hence, a for loop's else part runs if no break occurs.

Here is an example to illustrate this.

In [23]:
digits = [0, 1, 5]

for i in digits:
    print(i)
else:
    print("No items left.")

0
1
5
No items left.


# While loop in Python

* The while loop in Python is used to iterate over a block of code as long as the test expression (condition) is true.

* We generally use this loop when we don't know beforehand, the number of times to iterate.

In [24]:
# Program to add natural
# numbers upto 
# sum = 1+2+3+...+n

# To take input from the user,
# n = int(input("Enter n: "))

n = 10

# initialize sum and counter
sum = 0
i = 1

while i <= n:
    sum = sum + i
    i = i+1    # update counter

# print the sum
print("The sum is", sum)



The sum is 55


In [25]:
# Example to illustrate
# the use of else statement
# with the while loop

counter = 0

while counter < 3:
    print("Inside loop")
    counter = counter + 1
else:
    print("Inside else")

Inside loop
Inside loop
Inside loop
Inside else


# Functions

Syntax :

`
def function_name(parameters):
	"""docstring"""
	statement(s)
`

* Keyword def marks the start of function header.
* A function name to uniquely identify it. 
* Function naming follows the same rules of writing identifiers in Python.
* Parameters (arguments) through which we pass values to a function. They are optional.
* A colon (:) to mark the end of function header.
* Optional documentation string (docstring) to describe what the function does.
* One or more valid python statements that make up the function body. Statements must have same indentation level (usually 4 spaces).
* An optional return statement to return a value from the function.

In [26]:
def greet(name):
	"""This function greets to
	the person passed in as
	parameter"""
	print("Hello, " + name + ". Good morning!")
    
    
def absolute_value(num):
	"""This function returns the absolute
	value of the entered number"""

	if num >= 0:
		return num
	else:
		return -num

# Output: 2
print(absolute_value(2))

# Output: 4
print(absolute_value(-4))

2
4


## Scope and Lifetime of variables

* Scope of a variable is the portion of a program where the variable is recognized. Parameters and variables defined inside a function is not visible from outside. Hence, they have a local scope.

* Lifetime of a variable is the period throughout which the variable exits in the memory. The lifetime of variables inside a function is as long as the function executes.

* They are destroyed once we return from the function. Hence, a function does not remember the value of a variable from its previous calls.

In [27]:
def my_func():
	x = 10
	print("Value inside function:",x)

x = 20
my_func()
print("Value outside function:",x)

Value inside function: 10
Value outside function: 20


## Python Keyword Arguments

When we call a function with some values, these values get assigned to the arguments according to their position.

Functions can have 
* positional arguments 
* keyword arguments 
* a fixed number of arguments
* variable number of arguments

we can mix positional arguments with keyword arguments during a function call. But we must keep in mind that **keyword arguments must follow positional arguments**.


In [28]:
# function with fixed number of arguments
def greet(name,msg):
   """This function greets to
   the person with the provided message"""
   print("Hello",name + ', ' + msg)

greet("Monica","Good morning!")




Hello Monica, Good morning!


In [29]:
def greet(name, msg = "Good morning!"):
   """
   This function greets to
   the person with the
   provided message.

   If message is not provided,
   it defaults to "Good
   morning!"
   """

   print("Hello",name + ', ' + msg)

greet("Ayşe")
greet("Mehmet","How do you do?")
# 1 positional, 1 keyword argument
greet("Ali",msg = "Are you allright?") 

Hello Ayşe, Good morning!
Hello Mehmet, How do you do?
Hello Ali, Are you allright?


## Understanding *args and **kwargs

* In Python, the single-asterisk form of *args can be used as a parameter to send a non-keyworded variable-length argument list to functions. 
* It is worth noting that the asterisk (*) is the important element here, as the word args is the established conventional idiom, though it is not enforced by the language.

* The double asterisk form of **kwargs is used to pass a keyworded, variable-length argument dictionary to a function. Again, the two asterisks (**) are the important element here, as the word kwargs is conventionally used, though not enforced by the language.

* Like *args, **kwargs can take however many arguments you would like to supply to it. However, **kwargs differs from *args in that you will need to assign keywords.


In [30]:
def multiply(*args):
    z = 1
    for num in args:
        z *= num
    print(z)

multiply(4, 5)
multiply(10, 9)
multiply(2, 3, 4)
multiply(3, 5, 10, 6)

20
90
24
900


In [31]:
def print_values(**kwargs):
    for key, value in kwargs.items():
        print("The value of {} is {}".format(key, value))

print_values(my_name="Sammy", your_name="Casey")

The value of my_name is Sammy
The value of your_name is Casey


# Python Exception Handling - Try, Except and Finally

* Python has many built-in exceptions which forces your program to output an error when something in it goes wrong.

* When these exceptions occur, it causes the current process to stop and passes it to the calling process until it is handled. If not handled, our program will crash.

* In Python, exceptions can be handled using a try statement.

* A critical operation which can raise exception is placed inside the try clause and the code that handles exception is written in except clause.

In [32]:
# import module sys to get the type of exception
import sys

randomList = ['a', 0, 2]

for entry in randomList:
    try:
        print("The entry is", entry)
        r = 1/int(entry)
        break
    except:
        print("Oops!",sys.exc_info()[0],"occured.")
        print("Next entry.")
        print()
print("The reciprocal of",entry,"is",r)

The entry is a
Oops! <class 'ValueError'> occured.
Next entry.

The entry is 0
Oops! <class 'ZeroDivisionError'> occured.
Next entry.

The entry is 2
The reciprocal of 2 is 0.5


## Catching Specific Exceptions and raising exceptions


In [33]:
# catching specific exceptions

try:
   # do something
   pass

except ValueError:
   # handle ValueError exception
   pass

except (TypeError, ZeroDivisionError):
   # handle multiple exceptions
   # TypeError and ZeroDivisionError
   pass

except:
   # handle all other exceptions
   pass


## -----------------------
# raising an exception
try:
     a = int(input("Enter a positive integer: "))
     if a <= 0:
         raise ValueError("That is not a positive number!")
except ValueError as ve:
     print(ve)

Enter a positive integer: -3
That is not a positive number!


# Python Object Oriented Programming

Python is a **multi-paradigm** programming language. Meaning, it supports different programming approach.

One of the popular approach to solve a programming problem is by creating objects. This is known as Object-Oriented Programming (OOP).

An object has two characteristics:

* attributes
* behavior

In Python, the concept of OOP follows some basic principles:
* **inheritance**: A process of using details from a new class without modifying existing class.
* **encapsulation**: Hiding the private details of a class from other objects.
* **polymorphism**: A concept of using common operation in different ways for different data input.


In [34]:
class Parrot:

    # class attribute
    species = "bird"

    # instance attribute
    def __init__(self, name, age):
        self.name = name
        self.age = age

# instantiate the Parrot class
blu = Parrot("Blu", 10)
woo = Parrot("Woo", 15)

# access the class attributes
print("Blu is a {}".format(blu.__class__.species))
print("Woo is also a {}".format(woo.__class__.species))

# access the instance attributes
print("{} is {} years old".format( blu.name, blu.age))
print("{} is {} years old".format( woo.name, woo.age))

Blu is a bird
Woo is also a bird
Blu is 10 years old
Woo is 15 years old


## Inheritance

Inheritance is a way of creating new class for using details of existing class without modifying it. The newly formed class is a derived class (or child class). Similarly, the existing class is a base class (or parent class).


In [35]:
# parent class
class Bird:
    
    def __init__(self):
        print("Bird is ready")

    def whoisThis(self):
        print("Bird")

    def swim(self):
        print("Swim faster")

# child class
class Penguin(Bird):

    def __init__(self):
        # call super() function
        super().__init__()
        print("Penguin is ready")

    def whoisThis(self):
        print("Penguin")

    def run(self):
        print("Run faster")

peggy = Penguin()
peggy.whoisThis()
peggy.swim()
peggy.run()

Bird is ready
Penguin is ready
Penguin
Swim faster
Run faster


## Encapsulation

Using OOP in Python, we can restrict access to methods and variables. This prevent data from direct modification which is called encapsulation. In Python, we denote private attribute using underscore as prefix i.e single “ _ “ or double “ __“.


In [36]:
class Computer:

    def __init__(self):
        self.__maxprice = 900

    def sell(self):
        print("Selling Price: {}".format(self.__maxprice))

    def setMaxPrice(self, price):
        self.__maxprice = price

c = Computer()
c.sell()

# change the price
c.__maxprice = 1000
c.sell()

# using setter function
c.setMaxPrice(1000)
c.sell()

Selling Price: 900
Selling Price: 900
Selling Price: 1000


## Polymorphism

Polymorphism is an ability (in OOP) to use common interface for multiple form (data types).

Suppose, we need to color a shape, there are multiple shape option (rectangle, square, circle). However we could use same method to color any shape. This concept is called Polymorphism.


In [39]:
class Parrot:

    def fly(self):
        print("Parrot can fly")
    
    def swim(self):
        print("Parrot can't swim")
class Penguin:

    def fly(self):
        print("Penguin can't fly")
    
    def swim(self):
        print("Penguin can swim")

# common interface
def flying_test(bird):
    bird.fly()

#instantiate objects
blu = Parrot()
peggy = Penguin()

# passing the object
flying_test(blu)
flying_test(peggy)

Parrot can fly
Penguin can't fly


## Lambda Expressions
* Small anonymous functions can be created with the lambda keyword. 

* This function returns the sum of its two arguments: `lambda a, b: a+b`. 
* Lambda functions can be used wherever function objects are required. 
* They are syntactically restricted to a single expression. Semantically, they are just syntactic sugar for a normal function definition. 
* Like nested function definitions, lambda functions can reference variables from the containing scope:

In [41]:
def make_incrementor(n):
     return lambda x: x + n

f = make_incrementor(42)
print(f)
print(f(0))
print(f(1))
print(f(3))

<function make_incrementor.<locals>.<lambda> at 0x0000020020049510>
42
43
45


In [42]:
# Program to filter out only the even items from a list
# filter(function, iterable) 
my_list = [1, 5, 4, 6, 8, 11, 3, 12]

new_list = list(filter(lambda x: (x%2 == 0) , my_list))

# Output: [4, 6, 8, 12]
print(new_list)

[4, 6, 8, 12]


In [43]:
'''Map applies a function to all the items in an input_list. Here is the blueprint:'''
# syntax:  map(function_to_apply, list_of_inputs)
colors = ["Goldenrod", "Purple", "Salmon", "Turquoise", "Cyan"]
print(list(map(lambda s: s.casefold(), colors)))

'''another example'''
items = [1, 2, 3, 4, 5]
squared = list(map(lambda x: x**2, items))
print(squared)

['goldenrod', 'purple', 'salmon', 'turquoise', 'cyan']
[1, 4, 9, 16, 25]


##  A Quick Introducton to Generators and `yield` keyword in Python

* **Generators** allow the programmer to specify that a function can be prematurely exited and then later re-entered at the point of last exit, enabling coroutines, meaning functions that alternate execution with each other. 
* The exit/re-entry points are marked by Python’s **`yield`** keyword.
* Each new call to the function causes a resumption of execution of the function at the point immediately following the last yield executed in that function. 
* As you will see below, that is exactly what we need for DES.

<h2><a id="difference" name="difference"></a>Differences between Generator function and a Normal function</h2>

<p>Here is how a generator function differs from a <a href="/python-programming/function" title="Python functions">normal function</a>.</p>

<ul>
	<li>Generator function contains one or more <code>yield</code> statement.</li>
	<li>When called, it returns an object (iterator) but does not start execution immediately.</li>
	<li>Methods like <code>__iter__()</code> and <code>__next__()</code> are implemented automatically. So we can iterate through the items using <code>next()</code>.</li>
	<li>Once the function yields, the function is paused and the control is transferred to the caller.</li>
	<li>Local variables and their states are remembered between successive calls.</li>
	<li>Finally, when the function terminates, <code>StopIteration</code> is raised automatically on further calls.</li>
</ul>


In [44]:
def myFunction(n):
    for i in range(n):
        return i

def myGenerator(n):
    while True:
        for i in range(n):
            yield i

t = myFunction(3)
print('t = ', t)
print("now printing generator output")
g = myGenerator(3)
print('g = ', g)
print(g.__next__())
print(g.__next__())
print(g.__next__())
print(g.__next__())
print("creating a new generator object")
g1 = myGenerator(2)
print(g1.__next__())
print(g1.__next__())
print(g1.__next__())
print(g1.__next__())


t =  0
now printing generator output
g =  <generator object myGenerator at 0x000002001FFECD58>
0
1
2
0
creating a new generator object
0
1
0
1


In [45]:
def reverse(data):
    for index in range(len(data)-1, -1, -1):
        yield data[index]

for char in reverse('program'):
     print(char)

m
a
r
g
o
r
p
